### PR TITLE
`mapParsedUrl` and `mapUrl`

### PR DESCRIPTION
**Pull request by OliverJAsh**
Depends on #4

### CODE REVIEW

File: src/tests.ts [hunk 1]
```diff
@@ -1,13 +1,57 @@
...
+
+assert.deepEqual(
+    pipeWith('https://foo.com/bar', url =>
+        mapUrl(({ parsedUrl }) => ({
+            ...parsedUrl,
+            pathname: '/foo',
+            query: { a: 'b' },
+        }))({ url }),
+    ),
+    'https://foo.com/foo?a=b',
```
Comment (OliverJAsh) [inline src/tests.ts:39]: Now you can run multiple operations on a URL all at once, e.g. the above as an alternative to running "replace pathname" and then "replace query".

File: src/tests.ts [hunk 2]
```diff
@@ -1,13 +1,57 @@
...
+    pipeWith('https://foo.com/bar', url =>
+        mapUrl(
+            pipe(
+                replacePathnameInParsedUrl({ newPathname: () => '/foo' }),
+                parsedUrl =>
+                    replaceQueryInParsedUrl({ newQuery: () => ({ a: 'b' }) })({ parsedUrl }),
+            ),
+        )({ url }),
+    ),
+    'https://foo.com/foo?a=b',
```
Comment (OliverJAsh) [inline src/tests.ts:52]: When running multiple operations on a URL all at once (via `mapUrl`), you can also compose update functions (a bit like transducers).

File: src/tests.ts (hunk 1 above)
Comment (karol-majewski) [reply src/tests.ts:39]: I was wondering if we need to solve that problem ourselves if the platform provides us with tools like `URL` and `URLSearchParams`. A minimal working implementation:

```ts
function updateURL(base: string, patch: Partial<URL>): string {
  return Object.assign(new URL(base), patch).href;
}
```

Usage:

```ts
updateURL('https://foo.com/bar', {
  pathname: '/foo',
  search: new URLSearchParams({ a: 'b' }).toString(),
}); // => "https://foo.com/foo?a=b'"
```

We could wrap these types to improve the functionality of course (for example, reject read-only properties), but such a solution is both simpler and has no dependencies. Throwing away the built-in solutions seems like a bit of a waste. Have you considered integrating them?

**Review by karol-majewski — APPROVED** [review]
If you're keen to extend our url-manipulation toolbelt with more helpers, I understand (and I won't stop you 😛), but at the same part of me thinks this is a solved problem — everything we do can also be achieved by the abstractions provided by the standard library.

If currying is what we're after, then we could wrap them in a way we see fit and still enjoy a dependency-free life.

Perhaps there are benefits to doing it ourselves I don't see right now (I'm sure there are), but given the increasing support for `URL` and `URLSearchParams`, it feels like the maintenance burden could be vastly reduced by delegating the problem to the platform.

File: src/tests.ts (hunk 1 above)
Comment (OliverJAsh) [reply src/tests.ts:39]: > Throwing away the built-in solutions seems like a bit of a waste. Have you considered integrating them?

We are using a built-in solution—Node's `url.format` and `url.parse`, which are the equivalent of `new URL` and `url.toString()` (respectively).

We have already discussed migrating away from the Node API and towards the WHATWG API (https://github.com/unsplash/unsplash-web/issues/4303). This library is just a very small abstraction (a handful of tiny wrapper functions) on top of that—what's used underneath (the "engine") doesn't really matter.

With that to one side, the remainder of the question seems to be about whether we could do away with the wrapper functions and always use something like your `updateURL` (which is the WHATWG equivalent of the current `mapUrl` function we have in this PR). We certainly could, but the wrapper functions do help in making things a bit more explicit, which is useful seeing as there are bunch of operations which we frequently run. For example:

```ts
mapUrl('https://foo.com/bar?a=b', url => ({ query: { ...url.query, c: 'd' } })); // => "https://foo.com/bar?a=b&c=d"
// vs
addQueryToUrl('https://foo.com/bar?a=b', { c: 'd' }); // => "https://foo.com/bar?a=b&c=d"
```

However when we're running multiple operations on a URL (e.g. "add query" _and_ "replace pathname") then it probably makes more sense to use `mapUrl`.

WDYT?

**Comment by OliverJAsh** [comment]
@karol-majewski I replied here: https://github.com/unsplash/url-transformers/pull/8#discussion_r358234798

Note this PR depends on #4, which needs approval separately. Would you mind taking a look?

File: src/tests.ts (hunk 1 above)
Comment (karol-majewski) [reply src/tests.ts:39]: > We are using a built-in solution

When I said _a built-in solution_, what I had in mind was the standard library available in the browser. The reasons why this is important:

1. Standard library helpers will _always_ be backward compatible.
1. No dependencies added.

The approach used by `URL` and `URLSearchParams` — converting primitives to objects — is also more flexible, because you get to inspect and modify all parts of your URL without having to import more helpers. Normally, that wouldn't always be desirable (because of tree shaking), but in this case, there is nothing to tree-shake — for the majority of users, these helpers will be built-in.

If the built-in helpers are missing something (like a `map` method or support for relative paths), we could extend them.

```ts
class URI extends URL {
  map() { /* ... */}
  relative() { /* ... */}
}
```

```ts
new URI('https://unsplash.com/photos/nqLMg7DHLqU/info?foo=bar#id')
  .map(uri => new URI({ ...uri, pathname: 'dogs' }))
  .relative()
  .toString();
```

> However when we're running multiple operations on a URL (e.g. "add query" _and_ "replace pathname") then it probably makes more sense to use `mapUrl`.

Precisely! When you say "URL", most people will think of the string representation of a URL, instead of taking it for what it is (a key-value mapping). And just like with any other object, it's not uncommon to modify more than one property at a time and use conditionals.

For example, if you want to change both `search` and `pathname`, it will be easier for you to use `mapUrl` instead of two specialized helpers.

If `mapUrl` gets the job done for two parts of your URL, it would do fine for a single part as well. In that light, why shouldn't one just use `mapUrl` in every scenario? What would be the incentive to use a more specialized helper instead of `mapUrl` that works everywhere? And, if `mapUrl` offers the same power `URI#map` described above does, then how is it better?

**TL;DR I like `mapUrl`, I would like to use it everywhere, but then it would be just like using the built-in `URL`, so why not go straight for  `URL` instead.**

Comment (OliverJAsh) [reply src/tests.ts:39]: > What would be the incentive to use a more specialized helper instead of mapUrl that works everywhere?

Some examples:

- `appendPathname` carefully handles trailing slashes when merging:
  ```ts
  assert.strictEqual(appendPathnameToUrl({ url: '/foo' })({ pathnameToAppend: '/bar' }), '/foo/bar');
  assert.strictEqual(appendPathnameToUrl({ url: '/foo/' })({ pathnameToAppend: '/bar' }), '/foo/bar');
  ```
- `addQuery` handles merging of query objects

With `mapUrl` you'd have to remember to do these things each time you use it. They're small things, but we do them often enough to warrant small wrappers—the same reasons we might choose to compose any function.

We may well be able to get rid of some wrappers which aren't doing much to help, e.g. `replaceHashInUrl`. Once this PR is out, we can begin using `mapUrl` and try to migrate to using that instead where it makes sense to do so. WDYT?

> And, if mapUrl offers the same power URI#map described above does, then how is it better?

We can use `mapUrl` right now—the same is not true for `URI#map` without further work (https://github.com/unsplash/unsplash-web/issues/4303), but even then it's conditional on us finding a way to make relative paths work.

> support for relative paths

Btw I'm not so sure this is an easy thing for us to add, but we'll cross that bridge when we come to it. I started experimenting here: #10.

**Comment by karol-majewski** [comment]
@OliverJAsh sure, I'll have a look.
